File: core/src/main/java/com/stevenlagoy/presidency/map/travel/route/Airport.java
package com.stevenlagoy.presidency.map.travel.route;

import com.stevenlagoy.jsonic.JSONObject;
import com.stevenlagoy.jsonic.Jsonic;
import com.stevenlagoy.presidency.map.MapManager;
import com.stevenlagoy.presidency.map.Municipality;

public class Airport implements Jsonic<Airport> {

    // STATIC VARIABLES
    // ---------------------------------------------------------------------------

    public static enum AirportSize {
        LARGE,
        MEDIUM,
        SMALL;

        public static AirportSize fromString(String str) {
            str = str.toUpperCase().replaceAll("\\s+", "").trim();
            return switch (str) {
                case "LARGE" -> LARGE;
                case "L" -> LARGE;
                case "MEDIUM" -> MEDIUM;
                case "M" -> MEDIUM;
                case "SMALL" -> SMALL;
                case "S" -> SMALL;
                default -> null;
            };
        }
    }

    // INSTANCE VARIABLES
    // -------------------------------------------------------------------------

    private String fullName;
    private String commonName;
    private String IATA;
    private Municipality location;
    private AirportSize size;
    private int enplanement;

    // CONSTRUCTORS
    // -------------------------------------------------------------------------------

    public Airport() {
        fullName = "";
        commonName = "";
        IATA = "";
        location = null;
        size = null;
        enplanement = 0;
    }

    public Airport(String fullName, String commonName, String IATA, Municipality location, AirportSize size,
                   int enplanement) {
        this.fullName = fullName;
        this.commonName = commonName;
        this.IATA = IATA;
        this.location = location;
        this.size = size;
        this.enplanement = enplanement;
    }

    public Airport(MapManager mm, String fullName, String commonName, String IATA, String locationName, String size,
                   int enplanement) {
    }

    public String getFullName() {
        return fullName;
    }

    public String getCommonName() {
        return commonName;
    }

    public String getIATA() {
        return IATA;
    }

    public Municipality getMunicipality() {
        return location;
    }

    public AirportSize getSize() {
        return size;
    }

    public int getEnplanement() {
        return enplanement;
    }

    // REPRESENTATION METHODS
    // ---------------------------------------------------------------------

    @Override
    public JSONObject toJson() {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'toJson'");
    }

    @Override
    public Airport fromJson(JSONObject json) {
        // TODO Auto-generated method stub
        throw new UnsupportedOperationException("Unimplemented method 'fromJson'");
    }
}

File: core/src/main/kotlin/com/stevenlagoy/presidency/map/travel/route/Airport.kt
package com.stevenlagoy.presidency.map.travel.route

import com.stevenlagoy.jsonic.JSONObject
import com.stevenlagoy.presidency.core.Engine
import com.stevenlagoy.presidency.map.entities.MapEntity

class Airport(
    engine: Engine,
    name: String = "",
    commonName: String = "",
    IATA: String = "",
    _location: MapEntity? = null,
    size: AirportSize = AirportSize.SMALL,
    enplanement: Int = 0,
    connections: List<MapEntity> = listOf(),
) : Route(engine, name, connections) {

    var commonName = commonName
        internal set

    var IATA = IATA
        internal set

    lateinit var location: MapEntity
        internal set

    var size = size
        internal set

    var enplanement = enplanement
        internal set

    init {
        if (_location != null) location = _location
    }

    constructor(engine: Engine, json: JSONObject) : this(engine) {
        fromJson(json)
    }

    override var connections: List<MapEntity> = connections
//        get() = if (location?.state?.censusDivision == null) return emptyList() else when (size) {
//            AirportSize.LARGE ->
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirports(AirportSize.LARGE).map { it.location } +
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirportsInRegion(location!!.state.censusDivision!!.censusRegion, AirportSize.MEDIUM).map { it.location }
//            AirportSize.MEDIUM ->
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirportsInRegion(location!!.state.censusDivision!!.censusRegion, AirportSize.LARGE).map { it.location } +
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirportsInDivision(location!!.state.censusDivision!!, AirportSize.SMALL).map { it.location }
//            AirportSize.SMALL ->
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirportsInDivision(location!!.state.censusDivision!!, AirportSize.MEDIUM).map { it.location } +
//                engine.MAP_MANAGER.ROUTE_MANAGER.getAirportsInDivision(location!!.state.censusDivision!!, AirportSize.SMALL).map { it.location }
//        }.filterNotNull()

    override fun toJson() = JSONObject(IATA, listOf(
        JSONObject("fullName", name),
        JSONObject("commonName", commonName),
        JSONObject("IATA", IATA),
        JSONObject("location", location.name),
        JSONObject("size", size.name),
        JSONObject("enplanement", enplanement),
    ))

    override fun fromJson(json: JSONObject) = this.apply {
        super.fromJson(json)
        commonName = json.requireString("commonName")
        IATA = json.requireString("IATA")
        location = engine.MAP_MANAGER.matchPlace(json.requireString("location")).orElseThrow { IllegalArgumentException("Could not find location ${json.requireString("location")} for airport $name") }
        size = AirportSize.valueOf(json.requireString("size").uppercase().replace(Regex("[^A-Z]"), "_"))
        enplanement = json.requireInt("enplanement")
    }

    enum class AirportSize {
        LARGE, MEDIUM, SMALL;
    }

}
